File: 3286. Find a Safe Walk Through a Grid.py
from typing import List
from collections import deque

class Solution:
    def findSafeWalk(self, grid: List[List[int]], health: int) -> bool:
        n = len(grid)
        m = len(grid[0])

        score = [[-1] * m for _ in range(n)]

        start_life = health - grid[0][0]
        if start_life < 1:
            return False

        score[0][0] = start_life

        q = deque()
        q.append((start_life, 0, 0))

        dx = [0, 0, 1, -1]
        dy = [1, -1, 0, 0]

        while q:
            life, x, y = q.popleft()

            if x == n - 1 and y == m - 1:
                return True

            for i in range(4):
                new_x = x + dx[i]
                new_y = y + dy[i]

                if 0 <= new_x < n and 0 <= new_y < m:
                    new_life = life - grid[new_x][new_y]

                    if new_life >= 1 and score[new_x][new_y] < new_life:
                        score[new_x][new_y] = new_life
                        q.append((new_life, new_x, new_y))

        return False

File: 3286. Find a Safe Walk Through a Grid.cpp
class Solution {
public:
    bool findSafeWalk(vector<vector<int>>& grid, int health) {
        int n = grid.size();
        int m = grid[0].size();

        vector<vector<int>> score(n, vector<int>(m, -1));

        int startLife = health - grid[0][0];
        if (startLife < 1)
            return false;

        score[0][0] = startLife;

        queue<pair<int, pair<int, int>>> q;
        q.push({startLife, {0, 0}});

        int dx[4] = {0, 0, 1, -1};
        int dy[4] = {1, -1, 0, 0};

        while (!q.empty()) {
            int life = q.front().first;
            int x = q.front().second.first;
            int y = q.front().second.second;
            q.pop();

            if (x == n - 1 && y == m - 1)
                return true;

            for (int i = 0; i < 4; i++) {
                int newX = x + dx[i];
                int newY = y + dy[i];

                if (newX >= 0 && newX < n &&
                    newY >= 0 && newY < m) {

                    int newLife = life - grid[newX][newY];

                    if (newLife >= 1 && score[newX][newY] < newLife) {
                        score[newX][newY] = newLife;
                        q.push({newLife, {newX, newY}});
                    }
                }
            }
        }

        return false;
    }
};